Kamping doesn't compile with C++20

When using C++ 20 with gcc, I get the following error message:


```
In file included from [...]/kamping/include/kamping/comm_helper/is_same_on_all_ranks.hpp:22,
                 from [...]/kamping/include/kamping/collectives/alltoall.hpp:25,
                 from [...]/kamping/examples/usage/alltoall_example.cpp:22:
[...]/kamping/include/kamping/mpi_ops.hpp:462:84: error: expected ‘)’ before ‘const’
  462 |     UserOperationPtrWrapper<is_commutative>(UserOperationPtrWrapper<is_commutative> const&) = delete;
      |                                            ~                                       ^~~~~~
      |                                                                                    )
[...]/kamping/include/kamping/mpi_ops.hpp:464:84: error: expected ‘)’ before ‘&&’ token
  464 |     UserOperationPtrWrapper<is_commutative>(UserOperationPtrWrapper<is_commutative>&& other_op) {
      |                                            ~                                       ^~
      |                                                                                    )
```

--- a/examples/applications/top_k_reduce_example.cpp
+++ b/examples/applications/top_k_reduce_example.cpp
@@ -15,7 +15,7 @@
 template <size_t K, typename ValueType>
 class TopK {
 public:
-    explicit TopK() {}
+    explicit TopK() noexcept {}
     ValueType& operator[](size_t i) {
         return elements[i];
     }

--- a/include/kamping/mpi_ops.hpp
+++ b/include/kamping/mpi_ops.hpp
@@ -459,9 +459,9 @@ public:
         return *this;
     }
 
-    UserOperationPtrWrapper<is_commutative>(UserOperationPtrWrapper<is_commutative> const&) = delete;
+    UserOperationPtrWrapper(UserOperationPtrWrapper<is_commutative> const&) = delete;
     /// @brief move constructor
-    UserOperationPtrWrapper<is_commutative>(UserOperationPtrWrapper<is_commutative>&& other_op) {
+    UserOperationPtrWrapper(UserOperationPtrWrapper<is_commutative>&& other_op) {
         this->_mpi_op   = other_op._mpi_op;
         this->_no_op    = other_op._no_op;
         other_op._no_op = true;

--- a/tests/collectives/mpi_allreduce_test.cpp
+++ b/tests/collectives/mpi_allreduce_test.cpp
@@ -50,13 +50,13 @@ TEST(AllreduceTest, allreduce_builtin_op_on_non_builtin_type) {
     Communicator comm;
 
     struct MyInt {
-        MyInt() : _value(0) {}
-        MyInt(int value) : _value(value) {}
+        MyInt() noexcept : _value(0) {}
+        MyInt(int value) noexcept : _value(value) {}
         int _value;
         int operator+(MyInt const& rhs) const noexcept {
             return this->_value + rhs._value;
         }
-        bool operator==(MyInt const& rhs) const {
+        bool operator==(MyInt const& rhs) const noexcept {
             return this->_value == rhs._value;
         }
     };

--- a/tests/collectives/mpi_exscan_test.cpp
+++ b/tests/collectives/mpi_exscan_test.cpp
@@ -148,13 +148,13 @@ TEST(ExscanTest, builtin_op_on_non_builtin_type) {
     Communicator comm;
 
     struct MyInt {
-        MyInt() : _value(0) {}
-        MyInt(int value) : _value(value) {}
+        MyInt() noexcept : _value(0) {}
+        MyInt(int value) noexcept : _value(value) {}
         int _value;
         int operator+(MyInt const& rhs) const noexcept {
             return this->_value + rhs._value;
         }
-        bool operator==(MyInt const& rhs) const {
+        bool operator==(MyInt const& rhs) const noexcept {
             return this->_value == rhs._value;
         }
     };
@@ -172,13 +172,13 @@ TEST(ExscanTest, identity_not_auto_deducible_and_no_values_on_rank_0_provided) {
     Communicator comm;
 
     struct MyInt {
-        MyInt() : _value(0) {}
-        MyInt(int value) : _value(value) {}
+        MyInt() noexcept : _value(0) {}
+        MyInt(int value) noexcept : _value(value) {}
         int _value;
         int operator+(MyInt const& rhs) const noexcept {
             return this->_value + rhs._value;
         }
-        bool operator==(MyInt const& rhs) const {
+        bool operator==(MyInt const& rhs) const noexcept {
             return this->_value == rhs._value;
         }
     };

--- a/tests/collectives/mpi_reduce_test.cpp
+++ b/tests/collectives/mpi_reduce_test.cpp
@@ -229,13 +229,13 @@ TEST(ReduceTest, reduce_builtin_op_on_non_builtin_type) {
     Communicator comm;
 
     struct MyInt {
-        MyInt() : _value(0) {}
-        MyInt(int value) : _value(value) {}
+        MyInt() noexcept : _value(0) {}
+        MyInt(int value) noexcept : _value(value) {}
         int _value;
         int operator+(MyInt const& rhs) const noexcept {
             return this->_value + rhs._value;
         }
-        bool operator==(MyInt const& rhs) const {
+        bool operator==(MyInt const& rhs) const noexcept {
             return this->_value == rhs._value;
         }
     };

--- a/tests/collectives/mpi_scan_test.cpp
+++ b/tests/collectives/mpi_scan_test.cpp
@@ -133,13 +133,13 @@ TEST(ScanTest, scan_builtin_op_on_non_builtin_type) {
     Communicator comm;
 
     struct MyInt {
-        MyInt() : _value(0) {}
-        MyInt(int value) : _value(value) {}
+        MyInt() noexcept : _value(0) {}
+        MyInt(int value) noexcept : _value(value) {}
         int _value;
         int operator+(MyInt const& rhs) const noexcept {
             return this->_value + rhs._value;
         }
-        bool operator==(MyInt const& rhs) const {
+        bool operator==(MyInt const& rhs) const noexcept {
             return this->_value == rhs._value;
         }
     };

--- a/tests/comm_helper/mpi_communicator_is_same_on_all_ranks_compilation_failure_test.cpp
+++ b/tests/comm_helper/mpi_communicator_is_same_on_all_ranks_compilation_failure_test.cpp
@@ -20,7 +20,8 @@ int main(int /*argc*/, char** /*argv*/) {
 
     [[maybe_unused]] Communicator comm;
     using NotAPod = std::vector<int>;
-    static_assert(!std::is_pod_v<NotAPod>);
+    // std::is_pod_v is deprecated in C++20
+    // static_assert(!std::is_pod_v<NotAPod>);
     [[maybe_unused]] int value = 0;
 
 #if defined(VALUE_IS_A_POINTER)

